src/peft/tuners/lora/layer.py

```diff
@@ -239,8 +240,16 @@ def dora_init(self, adapter_name: str) -> None:
         dora_layer = DoraLinearLayer(fan_in_fan_out=getattr(self, "fan_in_fan_out", False))
         lora_A = self.lora_A[adapter_name].weight
         lora_B = self.lora_B[adapter_name].weight
+        place_on_cpu = self.ephemeral_transfers and (lora_A.device.type == "cpu" or lora_B.device.type == "cpu")
+        if self.ephemeral_transfers:
+            if lora_A.device.type == "cuda":
+                lora_B = lora_B.to(lora_A.device)
+            else:
+                if lora_B.device.type != "cuda":
+                    lora_B = lora_B.to("cuda")
```

Comment:
what about the multi-GPU case here ? What would happen if the model is split across multiple GPUs, I suspect we might face a device mismatch ..

Reply:
This code is explicitly written to support multi GPU. Let's look at how the possible combinations will play out:

Let's denote chronological conditions as follows:
1. `A` is whether lora_A is on a "cuda" device; it infers that lora_B is placed on the same device (including device number) as lora_A. (L257-258)
2. `B` is whether lora_B is **not** on  a "cuda" device. It infers that lora_B is placed on an arbitrary cuda device (L260-261)
3. `not A` infers that lora_A is copied to the same device (including device number) as lora B

1. lora_A, lora_B on CPU: `not A → B` (cuda:A, cuda:A)
2. lora_A on CPU, lora_B on cuda:B: `not A → not B` (cuda:B, cuda:B)
3. lora_A on cuda:C, lora_B on cuda:D: `A` (cuda:C, cuda:C)

Reply:
OK, I think I got tests in place that check all possible combinations on a multi GPU setup.